contracts/core/modules/collect/ReFiCollectModule.sol

```diff
@@ -0,0 +1,240 @@
+        uint256 treasuryAmount = (amount * treasuryFee) / BPS_MAX;
+        uint256 adjustedAmount = amount - treasuryAmount;
+        address recipient = _dataByPublicationByProfile[profileId][pubId].recipient;
+        IERC20(currency).safeTransferFrom(collector, recipient, adjustedAmount);
+        
+        _buyToken(collector, profileId, pubId, data);
+    }
+
+    function _processCollectWithReferral(
+        uint256 referrerProfileId,
+        address collector,
+        uint256 profileId,
+        uint256 pubId,
+        bytes calldata data
+    ) internal {
+        uint256 amount = _dataByPublicationByProfile[profileId][pubId].amount;
+        address currency = _dataByPublicationByProfile[profileId][pubId].currency;
+        _validateDataIsExpected(data, currency, amount);
+
+        uint256 referralFee = _dataByPublicationByProfile[profileId][pubId].referralFee;
+        address treasury;
+        uint256 treasuryAmount;
+
+        // Avoids stack too deep
+        {
+            uint16 treasuryFee;
+            (treasury, treasuryFee) = _treasuryData();
+            treasuryAmount = (amount * treasuryFee) / BPS_MAX;
+        }
+
+        uint256 adjustedAmount = amount - treasuryAmount;
+
+        if (referralFee != 0) {
+            // The reason we levy the referral fee on the adjusted amount is so that referral fees
+            // don't bypass the treasury fee, in essence referrals pay their fair share to the treasury.
+            uint256 referralAmount = (adjustedAmount * referralFee) / BPS_MAX;
+            adjustedAmount = adjustedAmount - referralAmount;
+
+            address referralRecipient = IERC721(HUB).ownerOf(referrerProfileId);
+
+            IERC20(currency).safeTransferFrom(collector, referralRecipient, referralAmount);
+        }
+
+        address recipient = _dataByPublicationByProfile[profileId][pubId].recipient;
+        IERC20(currency).safeTransferFrom(collector, recipient, adjustedAmount);
+
+        _buyToken(collector, profileId, pubId, data);
+    }
+
+    /*
+    * @dev data is already verified!
+    */
+    function _buyToken(
+        address collector,
+        uint256 profileId,
+        uint256 pubId,
+        bytes calldata data
+    ) internal {
+        (address currency, uint256 initialAmount, uint256 amountOutMin) = abi.decode(
+            data, (address, uint256, uint256)
+        );
+        address tokenToBuy = _dataByPublicationByProfile[profileId][pubId].tokenToBuy;
+        address router = _dataByPublicationByProfile[profileId][pubId].router;
+    
+        (address treasury, uint16 treasuryFee) = _treasuryData();
+        uint256 amount = (initialAmount * treasuryFee) / BPS_MAX;
+
+        // Transfer tokens to this contract
+        IERC20(currency).safeTransferFrom(collector, address(this), amount);
+
+        // Increase allowance
+        IERC20(currency).approve(router, amount);
+
+        // Swap
+        address[] memory path = new address[](2);
+        path[0] = currency;
```

Comment:
FYI this approach will fail when the user provides a whitelisted currency other than USDC, since there is no direct route from e.g. WETH to BCT

Reply:
Thanks - that's a great observation. Please see [35](https://github.com/aave/lens-protocol/pull/35).